File: 2020/04_passport_processing/aoc04.jl
# Passport Processing
#
# Advent of Code 2020, day 4
# Solution by Geir Arne Hjelle, 2020-12-12

module AOC04

using Pipe


struct Height
    value::Union{Int16,Missing}
    unit::String

    function Height(height)
        m = match(r"^(?<value>\d+)(?<unit>(cm|in))$", height)
        if isnothing(m)
            new(missing, height)
        else
            new(parse(Int, m[:value]), m[:unit])
        end
    end
end

Passport = @NamedTuple begin
    byr::Union{Int16,Nothing}
    iyr::Union{Int16,Nothing}
    eyr::Union{Int16,Nothing}
    hgt::Union{Height,Nothing}
    hcl::Union{String,Nothing}
    ecl::Union{String,Nothing}
    pid::Union{String,Nothing}
end

function parse_passport(line)
    fields = @pipe line |> split .|> split(_, ":") |> Dict
    Passport(
        [
            get(fields, "byr", "") |> byr -> tryparse(Int, byr),
            get(fields, "iyr", "") |> iyr -> tryparse(Int, iyr),
            get(fields, "eyr", "") |> eyr -> tryparse(Int, eyr),
            get(fields, "hgt", nothing) |> hgt -> isnothing(hgt) ? nothing : Height(hgt),
            get(fields, "hcl", nothing),
            get(fields, "ecl", nothing),
            get(fields, "pid", nothing),
        ]
    )
end

# Check that all required fields are present in the passport
function required_fields(passport)
    passport |> values .|> isnothing |> !any
end

# Check that the value of each field is valid
function valid_fields(passport)
    @pipe passport |> pairs |> collect |> map(v -> validate(v...), _) |> all
end

# Dispatch on each Passport key
function validate(key, value)
    isnothing(value) ? false : validate(Val(key), value)
end

# Birth Year: four digits; at least 1920 and at most 2002.
function validate(::Val{:byr}, value)
    1920 <= value <= 2002
end

# Issue Year: four digits; at least 2010 and at most 2020.
function validate(::Val{:iyr}, value)
    2010 <= value <= 2020
end

# Expiration Year: four digits; at least 2020 and at most 2030.
function validate(::Val{:eyr}, value)
    2020 <= value <= 2030
end

# Height: a number followed by either cm or in:
#   - If cm, the number must be at least 150 and at most 193.
#   - If in, the number must be at least 59 and at most 76.
function validate(::Val{:hgt}, value)
    if value.unit == "cm"
        150 <= value.value <= 193
    elseif value.unit == "in"
        59 <= value.value <= 76
    else
        false
    end
end

# Hair Color: a # followed by exactly six characters 0-9 or a-f.
function validate(::Val{:hcl}, value)
    occursin(r"^#[0-9a-f]{6}$", value)
end

# Eye Color: exactly one of: amb blu brn gry grn hzl oth.
function validate(::Val{:ecl}, value)
    value in Set(["amb", "blu", "brn", "gry", "grn", "hzl", "oth"])
end

# Passport ID: a nine-digit number, including leading zeroes.
function validate(::Val{:pid}, value)
    occursin(r"^[0-9]{9}$", value)
end

# Solve the problem for one file
function solve(input)
    # Parse input
    passport = @pipe input |> split(_, "\n\n")

    # Part 1
    part_1 = passport .|> parse_passport .|> required_fields |> sum

    # Part 2
    part_2 = passport .|> parse_passport .|> valid_fields |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: 2020/04_passport_processing/aoc04.py
"""Passport Processing

Advent of Code 2020, day 4
Solution by Geir Arne Hjelle, 2020-12-04
"""
# Standard library imports
import pathlib
import re
import sys

debug = print if "--debug" in sys.argv else lambda *_: None


def byr(value):
    """byr (Birth Year) - four digits; at least 1920 and at most 2002"""
    return 1920 <= int(value) <= 2002


def iyr(value):
    """iyr (Issue Year) - four digits; at least 2010 and at most 2020"""
    return 2010 <= int(value) <= 2020


def eyr(value):
    """eyr (Expiration Year) - four digits; at least 2020 and at most 2030"""
    return 2020 <= int(value) <= 2030


def hgt(value):
    """hgt (Height) - a number followed by either cm or in:

    If cm, the number must be at least 150 and at most 193.
    If in, the number must be at least 59 and at most 76.
    """
    match = re.match(r"^(?P<number>\d+)(?P<unit>cm|in)", value)
    if not match:
        return False

    parsed = match.groupdict()
    if parsed["unit"] == "cm":
        return 150 <= int(parsed["number"]) <= 193
    else:
        return 59 <= int(parsed["number"]) <= 76


def hcl(value):
    """hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f"""
    return bool(re.match(r"^#[0-9a-f]{6}$", value))


def ecl(value):
    """ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth."""
    return value in {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}


def pid(value):
    """pid (Passport ID) - a nine-digit number, including leading zeroes."""
    return bool(re.match("^[0-9]{9}$", value))


required = {
    "byr": byr,
    "iyr": iyr,
    "eyr": eyr,
    "hgt": hgt,
    "hcl": hcl,
    "ecl": ecl,
    "pid": pid,
}


def parse_passport(text):
    """Parse one passport"""
    return dict(f.split(":") for f in text.split())


def is_valid_1(passport):
    """Check that all required fields are present"""
    return not (set(required) - set(passport))


def is_valid_2(passport):
    """Check that all required fields satisfy extra conditions"""
    for field in required:
        if field not in passport:
            return False

        if not required[field](value=passport[field]):
            return False

    return True


def main(args):
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        print(f"\n{file_path}:")
        passports = [parse_passport(pp) for pp in file_path.read_text().split("\n\n")]

        # Part 1
        num_valid = sum(is_valid_1(pp) for pp in passports)
        print(f"{num_valid} passports have all required fields")

        # Part 2
        num_valid = sum(is_valid_2(pp) for pp in passports)
        print(f"{num_valid} passports satisfy all requirements")


if __name__ == "__main__":
    main(sys.argv[1:])
